lib/hyacinth/digital_object/resource_attributes.rb

```diff
@@ -11,51 +11,67 @@ def resource_attributes
       end
 
       def resource_imports
-        @resource_imports ||= resource_import_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @resource_imports ||= resource_import_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
       end
 
       def old_resources
-        @old_resources ||= resource_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @old_resources ||= resource_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
       end
 
       def deleted_resources
-        @deleted_resources ||= resource_attributes.map { |key| [key.to_s, nil] }.to_h.with_indifferent_access
+        @deleted_resources ||= resource_attributes.map { |key, _config| [key.to_s, nil] }.to_h.with_indifferent_access
+      end
+
+      def delete_resource(resource_name)
+        return unless self.resources[resource_name].present?
+
+        self.deleted_resources[resource_name] = self.resources[resource_name]
+        self.resources[resource_name] = nil
       end
 
       module ClassMethods
         def resource_attributes
-          @resource_attributes ||= Set.new # initialized here because it may not have been initialized in a subclass of the including class
+          @resource_attributes ||= {} # initialized here because it may not have been initialized in a subclass of the including class
           if self.superclass.respond_to?(:resource_attributes)
             @resource_attributes.merge(self.superclass.resource_attributes)
           else
             @resource_attributes
           end
         end
 
+        def resource_attribute_names
+          resource_attributes.keys.dup.freeze
+        end
+
         def resource_import_attributes
-          @resource_import_attributes ||= Set.new # initialized here because it may not have been initialized in a subclass of the including class
+          @resource_import_attributes ||= {} # initialized here because it may not have been initialized in a subclass of the including class
           if self.superclass.respond_to?(:resource_import_attributes)
             @resource_import_attributes.merge(self.superclass.resource_import_attributes)
           else
             @resource_import_attributes
           end
         end
 
-        def resource_attribute(resource_attribute_name)
-          @resource_attributes ||= Set.new
-          @resource_attributes << resource_attribute_name.to_sym
+        def resource_attribute(resource_attribute_name, config = {})
+          config ||= {}
+          @resource_attributes ||= {}
+          @resource_attributes[resource_attribute_name.to_sym] = config
 
-          @resource_import_attributes ||= Set.new
-          @resource_import_attributes << resource_attribute_name.to_sym
+          @resource_import_attributes ||= {}
+          @resource_import_attributes[resource_attribute_name.to_sym] = config
```

Comment:
I'm not sure that we benefit from also associating the resource_attribute config options with the resource_import_attributes.  This change may not be necessary?

Reply:
If I remember correctly there's a place where the import `merge` the resource attribute config (or I thought there was). Maybe not necessary after all?

Reply:
I don't remember any merging of configs for the resource_import_attributes (but I could be forgetting what you're referring to).  Do you mean merging related to the superclass?  That was a case of merging Sets of strings.